kong/runloop/handler.lua

```diff
@@ -721,25 +718,86 @@ return {
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or var.scheme
-        forwarded_host  = var.http_x_forwarded_host  or var.host
-        forwarded_port  = var.http_x_forwarded_port  or var.server_port
+        forwarded_proto = var.http_x_forwarded_proto              or scheme
+        forwarded_host  = var.http_x_forwarded_host               or host
+        forwarded_port  = tonumber(var.http_x_forwarded_port, 10) or port
 
       else
-        forwarded_proto = var.scheme
-        forwarded_host  = var.host
-        forwarded_port  = var.server_port
+        forwarded_proto = scheme
+        forwarded_host  = host
+        forwarded_port  = port
       end
 
-      local protocols = route.protocols
-      if (protocols and
-          protocols.https and not protocols.http and forwarded_proto ~= "https")
+      local protocols = route and route.protocols
+      if (protocols and protocols.https and not protocols.http and
+          forwarded_proto ~= "https")
       then
         ngx.header["connection"] = "Upgrade"
         ngx.header["upgrade"]    = "TLS/1.2, HTTP/1.1"
         return kong.response.exit(426, { message = "Please use HTTPS protocol" })
       end
 
+      -- Service-less HTTP Route
+      if not service then
+        local service_scheme
+        local service_host
+        local service_port
+
+        local http_host = var.http_host
+        if http_host then
+          http_host = lower(http_host)
+
+          -- Micro-optimization to make sure we use JITable functions in this
+          -- potentially hot path, e.g. this simple code doesn't warrant us
+          -- using something like socket.url.parse:
+          -- https://github.com/diegonehab/luasocket/blob/043e99771352aff47680b99f09b66a32f0cc3ef5/src/url.lua#L143-L192
+          local s = find(http_host, ":", 2, true)
+          if s then
+            local p = tonumber(sub(http_host, s + 1), 10)
+            if p and p >= 1 and p <= 65535 then
+              local h = sub(http_host, 1, s - 1)
+              if h == host then
+                -- assume there was no host on request-line
+                service_host = h
+                service_port = p
+              end
+            end
+
+          elseif http_host == host then
+            -- assume there was no host on request-line
+            service_host = host
+          end
+        end
+
+        if not service_host then
+          service_host = host == var.server_name and var.server_addr or host
+        end
+
+        if not service_port then
+          if scheme == "https" then
+            service_port = 443
+          elseif scheme == "http" then
+            service_port = 80
+          else
+            service_port = var.server_port
+          end
+        end
+
+        if service_port == 443 then
+          service_scheme = "https"
+        elseif service_port == 80 then
+          service_scheme = "http"
+        else
+          service_scheme = scheme
```

Comment:
Shouldn't we always set `service_scheme` to `scheme`?

Reply:
I am not sure, let's say you connect to Kong using `https` and have a `host` header of `Host: www.google.com:80`, Should we still connect to `www.google.com` on port `80` with `http` or `https`?

Reply:
I don't think we should add special cases here: simpler to just have a rule "scheme in is scheme out". 

Note that in future we plan on having a route property like "terminate_tls" (recall https://github.com/Kong/kong/blob/aa034cda1be49a05751cb2695b1b56bbbcfafd1e/kong/runloop/handler.lua#L627), using that would probably be the right place to let a user configure http vs https.